styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -126,44 +117,54 @@
   private final StateManager stateManager;
   private final Stats stats;
   private final int pollPodsIntervalSeconds;
-  private final ServiceAccountKeyManager serviceAccountKeyManager;
+  private final KubernetesGCPServiceAccountSecretManager serviceAccountSecretManager;
 
   private Watch watch;
 
-  private final Object secretMutationLock = new Object() {};
-
   KubernetesDockerRunner(NamespacedKubernetesClient client, StateManager stateManager, Stats stats,
-      ServiceAccountKeyManager serviceAccountKeyManager, int pollPodsIntervalSeconds) {
+      KubernetesGCPServiceAccountSecretManager serviceAccountSecretManager, int pollPodsIntervalSeconds) {
     this.stateManager = Objects.requireNonNull(stateManager);
     this.client = Objects.requireNonNull(client).inNamespace(NAMESPACE);
     this.stats = Objects.requireNonNull(stats);
     this.pollPodsIntervalSeconds = pollPodsIntervalSeconds;
-    this.serviceAccountKeyManager = serviceAccountKeyManager;
+    this.serviceAccountSecretManager = Objects.requireNonNull(serviceAccountSecretManager);
   }
 
   KubernetesDockerRunner(NamespacedKubernetesClient client, StateManager stateManager, Stats stats,
-      ServiceAccountKeyManager serviceAccountKeyManager) {
-    this(client, stateManager, stats, serviceAccountKeyManager, DEFAULT_POLL_PODS_INTERVAL_SECONDS);
+      KubernetesGCPServiceAccountSecretManager serviceAccountSecretManager) {
+    this(client, stateManager, stats, serviceAccountSecretManager, DEFAULT_POLL_PODS_INTERVAL_SECONDS);
   }
 
   @Override
   public String start(WorkflowInstance workflowInstance, RunSpec runSpec) throws IOException {
-    ensureSecrets(workflowInstance, runSpec);
+    final KubernetesSecretSpec secretSpec = ensureSecrets(workflowInstance, runSpec);
     try {
-      final Pod pod = client.pods().create(createPod(workflowInstance, runSpec));
+      final Pod pod = client.pods().create(createPod(workflowInstance, runSpec, secretSpec));
       return pod.getMetadata().getName();
     } catch (KubernetesClientException kce) {
       throw new IOException("Failed to create Kubernetes pod", kce);
     }
   }
 
-  private void ensureSecrets(WorkflowInstance workflowInstance, RunSpec runSpec) throws IOException {
-    ensureServiceAccountKeySecret(workflowInstance, runSpec);
-    ensureCustomSecret(workflowInstance, runSpec);
+  @Override
+  public void cleanup() throws IOException {
+    serviceAccountSecretManager.cleanup();
+  }
+
+  private KubernetesSecretSpec ensureSecrets(WorkflowInstance workflowInstance, RunSpec runSpec)
+      throws IOException {
+    return KubernetesSecretSpec.builder()
+        .customSecret(ensureCustomSecret(workflowInstance, runSpec))
+        .serviceAccountSecret(runSpec.serviceAccount().isPresent()
```

Comment:
Maybe `ensureServiceAccountKeySecret` should just propagate IOException as a runtime exception?